Plugins/Detray/src/DetrayMaterialConverter.cpp

```diff
@@ -0,0 +1,219 @@
+  // (1) homogeneous -> 1 x 1 bin grid with closed axes
+  auto homogeneousMaterial =
+      dynamic_cast<const HomogeneousSurfaceMaterial*>(&material);
+  if (homogeneousMaterial != nullptr) {
+    ACTS_VERBOSE(
+        "DetrayMaterialConverter: found homogeneous surface material, this "
+        "will be modelled as a 1x1 bin grid");
+    // A single bin entry: convert it and fill it
+    detray::io::material_slab_payload slab = convertMaterialSlab(
+        homogeneousMaterial->materialSlab(Vector3{0., 0., 0.}));
+    detray::io::grid_bin_payload<detray::io::material_slab_payload> slabBin{
+        {0, 0}, {slab}};
+    // Filling axes and bins
+    materialGrid.axes = homogenousAxesPayloads();
+    materialGrid.bins = {slabBin};
+    return materialGrid;
+  }
+  // (2) - binned material -> convert into grid structure
+  auto binnedMaterial = dynamic_cast<const BinnedSurfaceMaterial*>(&material);
+  if (binnedMaterial != nullptr) {
+    ACTS_VERBOSE("DetrayMaterialConverter: found binned surface material");
+
+    // BinUtility modifications
+    bool swapped = false;
+    // Get the bin utility (make a copy as we may modify it)
+    // Detray expects 2-dimensional grid, currently supported are
+    // x-y, r-phi, phi-z
+    BinUtility bUtility = binnedMaterial->binUtility();
+    // Turn the bin value into a 2D grid
+    if (bUtility.dimensions() == 1u) {
+      if (bUtility.binningData()[0u].binvalue == BinningValue::binR) {
+        // Turn to R-Phi
+        bUtility += BinUtility(1u, -M_PI, M_PI, closed, BinningValue::binR);
+      } else if (bUtility.binningData()[0u].binvalue == BinningValue::binZ) {
+        // Turn to Phi-Z - swap needed
+        BinUtility nbUtility(1u, -M_PI, M_PI, closed, BinningValue::binPhi);
+        nbUtility += bUtility;
+        bUtility = std::move(nbUtility);
+        swapped = true;
+      } else {
+        std::runtime_error("Unsupported binning for Detray");
+      }
+    } else if (bUtility.dimensions() == 2u &&
+               bUtility.binningData()[0u].binvalue == BinningValue::binZ &&
+               bUtility.binningData()[1u].binvalue == BinningValue::binPhi) {
+      BinUtility nbUtility(bUtility.binningData()[1u]);
+      nbUtility += bUtility.binningData()[0u];
+      bUtility = std::move(nbUtility);
+      swapped = true;
+    }
+
+    BinningValue bVal0 = bUtility.binningData()[0u].binvalue;
+    BinningValue bVal1 = bUtility.binningData()[1u].binvalue;
+
+    // Translate into grid index type
+    int gridIndexType = 0;
+    if (bVal0 == BinningValue::binR && bVal1 == BinningValue::binPhi) {
+      gridIndexType = 0;
+    } else if (bVal0 == BinningValue::binPhi && bVal1 == BinningValue::binZ) {
+      gridIndexType = 3;
+    } else if (bVal0 == BinningValue::binX && bVal1 == BinningValue::binY) {
+      gridIndexType = 2;
+    } else {
+      std::runtime_error(
+          "Unsupported binning for Detray, grid index type is set to " +
+          std::to_string(gridIndexType));
+    }
```

Comment:
@niermann999 - can you please check where in the payload I should fill `gridIndexType` ?

Reply:
Ok, I see, in the `toDetrayJson` I did actually the following:

```c++
nlohmann::json Acts::MaterialJsonConverter::toJsonDetray(
    const Acts::ISurfaceMaterial& surfaceMaterial, const Acts::Surface& surface,
    std::size_t surfaceIndex, std::map<std::size_t, std::size_t>& gridLink) {
  nlohmann::json jSurfaceMaterial;
```

So I remember the gridLink indices in this map ... in order to create a correct typed link, I guess I have to do the same?